### PR TITLE
feat: update default partition number and replica num in cluster mode

### LINKED ISSUE #1473: Update default partition number and replica num in cluster mode
The partition num and replica num will be set to one when table created currently.
replica num: min(tablet num, 3)
partiton num: 8

### PR DESCRIPTION
**Pull request by keyu813**
close #1473

### CODE REVIEW

File: hybridse/src/node/node_manager.cc [hunk 1]
```diff
@@ -443,19 +443,18 @@ UnaryExpr *NodeManager::MakeUnaryExprNode(ExprNode *left, FnOperator op) {
...
-    int partition_num = 1;
+    CreateTableOptions options;
     SqlNodeList partition_meta_list;
     if (nullptr != table_option_list) {
         for (auto node_ptr : table_option_list->GetList()) {
             if (nullptr != node_ptr) {
                 switch (node_ptr->GetType()) {
                     case kReplicaNum: {
-                        replica_num = dynamic_cast<ReplicaNumNode *>(node_ptr)->GetReplicaNum();
+                        options.replica_num = dynamic_cast<ReplicaNumNode *>(node_ptr)->GetReplicaNum();
```
Comment (dl239) [inline hybridse/src/node/node_manager.cc]: do not resolve these config here. Set `SqlNodeList *table_option_list` as `CreateStmt` member directly

File: hybridse/include/node/plan_node.h [hunk 2]
```diff
@@ -422,25 +420,15 @@ class CreatePlanNode : public LeafPlanNode {
...
-
-    void setReplicaNum(int replica_num) { replica_num_ = replica_num; }
-
-    int GetPartitionNum() const { return partition_num_; }
-
-    void setPartitionNum(int partition_num) { partition_num_ = partition_num; }
-
-    NodePointVector &GetDistributionList() { return distribution_list_; }
-    void SetDistributionList(const NodePointVector &distribution_list) { distribution_list_ = distribution_list; }
+    NodePointVector &GetTableOptionList() { return table_option_list_; }
```
Comment (dl239) [inline hybridse/include/node/plan_node.h]: Label const funciton

File: hybridse/include/node/sql_node.h [hunk 3]
```diff
@@ -1819,12 +1815,8 @@ class CreateStmt : public SqlNode {
 
     bool GetOpIfNotExist() const { return op_if_not_exist_; }
 
-    int GetReplicaNum() const { return replica_num_; }
-
-    int GetPartitionNum() const { return partition_num_; }
-
-    NodePointVector &GetDistributionList() { return distribution_list_; }
-    const NodePointVector &GetDistributionList() const { return distribution_list_; }
+    NodePointVector &GetTableOptionList() { return table_option_list_; }
```
Comment (dl239) [inline hybridse/include/node/sql_node.h]: const function

File: src/sdk/sql_cluster_router.cc [hunk 4]
```diff
@@ -1761,23 +1761,24 @@ base::Status SQLClusterRouter::HandleSQLCreateTable(hybridse::node::CreatePlanNo
...
+    auto ns_client = cluster_sdk_->GetNsClient();
+    std::vector<::openmldb::client::TabletInfo> tablets;
+    std::string msg;
+    bool ok = ns_client->ShowTablet(tablets, msg);
+    if (!ok) {
+        return base::Status(base::ReturnCode::kCreateTableFailedOnTablet, "set default replica num failed");
     }
+    int tablets_size = tablets.size();
+    // set dafault value
+    int default_replica_num = std::min(tablets_size, 3);
```
Comment (dl239) [inline src/sdk/sql_cluster_router.cc]: use FLAGS_replica_num

File: src/sdk/sql_cluster_router.cc [hunk 5]
```diff
@@ -1761,23 +1761,24 @@ base::Status SQLClusterRouter::HandleSQLCreateTable(hybridse::node::CreatePlanNo
...
-                                "Fail to create table with the distribution configuration in standalone mode");
-        }
+    auto ns_client = cluster_sdk_->GetNsClient();
+    std::vector<::openmldb::client::TabletInfo> tablets;
+    std::string msg;
+    bool ok = ns_client->ShowTablet(tablets, msg);
+    if (!ok) {
+        return base::Status(base::ReturnCode::kCreateTableFailedOnTablet, "set default replica num failed");
     }
+    int tablets_size = tablets.size();
```
Comment (dl239) [inline src/sdk/sql_cluster_router.cc]: can get all tablets by `cluster_sdk_->GetTablet()`

File: src/sdk/node_adapter.cc [hunk 6]
```diff
@@ -33,29 +33,66 @@ namespace openmldb::sdk {
...
-    // todo: change default setting
-    int replica_num = create_node->GetReplicaNum();
-    if (replica_num <= 0) {
-        status->msg = "CREATE common: replica_num should be bigger than 0";
-        status->code = hybridse::common::kUnsupportSql;
-        return false;
+    hybridse::node::NodePointVector distribution_list;
+
+    int replica_num;
+    int partition_num;
```
Comment (dl239) [inline src/sdk/node_adapter.cc]: init to 1 and delete the else branch blew

File: src/sdk/node_adapter.cc [hunk 7]
```diff
@@ -33,29 +33,66 @@ namespace openmldb::sdk {
...
+    // different default value for cluser mode and standalone mode
+    if (is_cluster_mode) {
+        replica_num = default_replica_num;
+        partition_num = 8;
```
Comment (dl239) [inline src/sdk/node_adapter.cc]: use FLAGS_partition_num

File: hybridse/include/node/sql_node.h (hunk 3 above)
Comment (keyu813) [reply hybridse/include/node/sql_node.h]: have overloaded an const function of `GetTableOptionList` in line 1819

File: hybridse/include/node/plan_node.h (hunk 2 above)
Comment (aceforeverd) [reply hybridse/include/node/plan_node.h]: `const` in both declaration start and end .

File: hybridse/include/node/plan_node.h [hunk 8]
```diff
@@ -422,25 +420,15 @@ class CreatePlanNode : public LeafPlanNode {
...
+    void SetTableOptionList(const NodePointVector &table_option_list) { table_option_list_ = table_option_list; }
```
Comment (aceforeverd) [inline hybridse/include/node/plan_node.h]: if not used, rm it

File: hybridse/include/node/sql_node.h [hunk 9]
```diff
@@ -1819,12 +1815,8 @@ class CreateStmt : public SqlNode {
...
+    const NodePointVector &GetTableOptionList() const { return table_option_list_; }
```
Comment (aceforeverd) [inline hybridse/include/node/sql_node.h:1862]: why's two getter method ?

File: hybridse/include/node/sql_node.h (hunk 3 above)
Comment (aceforeverd) [reply hybridse/include/node/sql_node.h]: misleading naming. if u want to change `table_option_list_` inside class
1. should return pointer instead
2. add a `mutable` prefix, e.g `NodePointVector* MutableTableOptionList`

File: src/sdk/node_adapter.cc [hunk 10]
```diff
@@ -33,29 +33,66 @@ namespace openmldb::sdk {
...
+        partition_num = 1;
     }
-    table->set_replica_num(static_cast<uint32_t>(replica_num));
-    int partition_num = create_node->GetPartitionNum();
-    if (partition_num <= 0) {
-        status->msg = "CREATE common: partition_num should be greater than 0";
-        status->code = hybridse::common::kUnsupportSql;
-        return false;
+    // resolve table_option_list
+    for (auto table_option : table_option_list) {
```
Comment (aceforeverd) [inline src/sdk/node_adapter.cc]: auto&

File: src/sdk/node_adapter.cc [hunk 11]
```diff
@@ -33,29 +33,66 @@ namespace openmldb::sdk {
...
+        switch (table_option->GetType()) {
+            case hybridse::node::kReplicaNum: {
+                replica_num = dynamic_cast<hybridse::node::ReplicaNumNode *>(table_option)->GetReplicaNum();
```
Comment (aceforeverd) [inline src/sdk/node_adapter.cc]: must `break` on branch end

File: src/sdk/node_adapter.cc [hunk 12]
```diff
@@ -33,29 +33,66 @@ namespace openmldb::sdk {
...
     table->set_name(table_name);
-    // todo: change default setting
-    int replica_num = create_node->GetReplicaNum();
-    if (replica_num <= 0) {
-        status->msg = "CREATE common: replica_num should be bigger than 0";
-        status->code = hybridse::common::kUnsupportSql;
-        return false;
+    hybridse::node::NodePointVector distribution_list;
+
+    int replica_num;
```
Comment (aceforeverd) [inline src/sdk/node_adapter.cc]: use uint32_t directly

File: hybridse/include/node/sql_node.h (hunk 3 above)
Comment (dl239) [reply hybridse/include/node/sql_node.h:1818]: if it is a overloaded function, add `override` label

Comment (keyu813) [reply hybridse/include/node/sql_node.h]: not a overloaded function, `GetTableOptionList` will const and not change `tablet_option_list_`, another function is `MutableTableOptionList` will change `tablet_option_list_`

File: src/sdk/sql_cluster_router.cc (hunk 5 above)
Comment (keyu813) [reply src/sdk/sql_cluster_router.cc]: cluster_sdk_->GetTablet() must pass `table_name` as parameter, it seems have no method to get all tablets in cluster
image

Comment (dl239) [reply src/sdk/sql_cluster_router.cc]: Add `GetAllTablet` method to `ClientManager`.

File: src/sdk/node_adapter.cc [hunk 13]
```diff
@@ -28,34 +28,82 @@
...
-    if (replica_num <= 0) {
-        status->msg = "CREATE common: replica_num should be bigger than 0";
-        status->code = hybridse::common::kUnsupportSql;
-        return false;
+    hybridse::node::NodePointVector distribution_list;
+
+    hybridse::node::StorageMode storage_mode = hybridse::node::kMemory;
+    // different default value for cluser and standalone mode
+    FLAGS_replica_num = 1;
+    FLAGS_partition_num = 1;
```
Comment (dl239) [inline src/sdk/node_adapter.cc]: don't modify FLAGS values

File: src/sdk/sql_cluster_router.cc [hunk 14]
```diff
@@ -1776,23 +1776,24 @@ base::Status SQLClusterRouter::HandleSQLCreateTable(hybridse::node::CreatePlanNo
...
+    // set dafault value
+    uint32_t default_replica_num = std::min(tablets_size, 3);
```
Comment (dl239) [inline src/sdk/sql_cluster_router.cc]: `uint32_t default_replica_num = std::min(tablets_size, FLAGS_replica_num);`

File: java/openmldb-jdbc/src/test/java/com/_4paradigm/openmldb/jdbc/SQLRouterSmokeTest.java [hunk 15]
```diff
@@ -119,50 +120,44 @@ public void testSmoke(SqlExecutor router) {
...
-            Assert.assertEquals("hi", rs1.getString(2));
-            Assert.assertEquals(1002, rs1.getLong(1));
+            List<Long> col1Insert = new ArrayList<>();
+            List<String> col2Insert = new ArrayList<>();
+            while (rs1.next()) {
+                col1Insert.add(rs1.getLong(1));
+                col2Insert.add(rs1.getString(2));
+            }
+            Collections.sort(col1Insert);
+            Collections.sort(col2Insert);
```
Comment (dl239) [inline java/openmldb-jdbc/src/test/java/com/_4paradigm/openmldb/jdbc/SQLRouterSmokeTest.java:131]: why sort

**Comment by aceforeverd** [comment]
rebase after #1636

**Review by dl239 — APPROVED** [review]

**Review by aceforeverd — APPROVED** [review]

**Pull request merged**